clang/test/SemaSYCL/allow-constexpr-recursion.cpp

```diff
@@ -0,0 +1,68 @@
+// RUN: %clang_cc1 -fsycl -fsycl-is-device -fcxx-exceptions -Wno-return-type -verify -fsyntax-only -std=c++20 -Werror=vla %s
```

Comment:
can this test be made c++17 as this is what we use today for SYCL?

Reply:
No, it cannot.  We're C++20 here because we need to test conditionally explicit constructors, which is a C++20 feature.

Reply:
I see. The problem is that recursion restriction relaxation will be tested in different compiler language level compared to what is used in SYCL, which reduces guarantees the test provides. But if you think it is still good enough - I'm OK.

Reply:
Do we need to test conditionally explicit constructors if it is a C++20 feature? Does it make sense to guard that part of the test and add a second run line for -std=c++17? Asking because of Konst's point about test guarantees above. I'll leave the final decision up to you.

Reply:
I can, but the logic here is not C++20/C++17 specific so I didn't particularly see the value.  Presumably one day SYCL will change to C++20, so having it in the test seems like a valuable thing to at least test.

Reply:
> Presumably one day SYCL will change to C++20, so having it in the test seems like a valuable thing to at least test.

While SYCL 2020 requires a C++17 compiler, it does not prevent using a more modern version. And that is what we do because C++20 allows simpler programs. So I guess all the tests for SYCL 1.2.1 should be run on all the C++ versions starting at C++11 and all the tests for SYCL 2020 should be run on all the C++ versions starting at C++17. I hope you have access to green electricity. :-)